src/test/java/org/example/bookstore/service/impl/BookServiceImplTest.java

```diff
@@ -0,0 +1,306 @@
+        BookDto expected = new BookDto();
+        expected.setId(BOOK_ID);
+        expected.setAuthor(UPDATED_AUTHOR);
+        expected.setIsbn(UPDATED_ISBN);
+        expected.setPrice(UPDATED_PRICE);
+        expected.setTitle(TITLE);
+        expected.setDescription(UPDATED_DESCRIPTION);
+        expected.setCoverImage(UPDATED_COVER_IMAGE);
+        expected.setCategoryIds(CATEGORY_IDS);
+        return expected;
+    }
+
+    @BeforeEach
+    void setUp() {
+        requestDto = createRequestDto();
+        book = createBook();
+    }
+
+    @Test
+    @DisplayName("Save valid book")
+    void save_ValidBook_ReturnsValidBookDto() {
+        BookDto expected = new BookDto();
+        expected.setId(BOOK_ID);
+        expected.setAuthor(AUTHOR);
+        expected.setIsbn(ISBN);
+        expected.setPrice(PRICE);
+        expected.setTitle(TITLE);
+        expected.setDescription(DESCRIPTION);
+        expected.setCoverImage(COVER_IMAGE);
+        expected.setCategoryIds(CATEGORY_IDS);
+
+        when(bookMapper.toModel(any(CreateBookRequestDto.class))).thenReturn(book);
+        when(bookRepository.save(any(Book.class))).thenReturn(book);
+        when(bookMapper.toDto(any(Book.class))).thenReturn(expected);
+        doNothing().when(categoryService).checkIfCategoriesExistsByIds(anySet());
+
+        BookDto actual = bookService.save(requestDto);
+
+        assertEquals(expected, actual);
+    }
+
+    @Test
+    @DisplayName("Save book with invalid category id")
+    void save_InvalidCategoryId_ThrowsException() {
+        doThrow(new EntityNotFoundException(EXPECTED_CATEGORY_NOT_FOUND_MESSAGE))
+                .when(categoryService).checkIfCategoriesExistsByIds(anySet());
+
+        EntityNotFoundException thrown =
+                assertThrows(EntityNotFoundException.class, () -> bookService.save(requestDto));
+
+        assertEquals(EXPECTED_CATEGORY_NOT_FOUND_MESSAGE, thrown.getMessage());
+    }
+
+    @Test
+    @DisplayName("Find all books with default pageable params")
+    void findAll_DefaultPageableParams_ReturnsAllBookDtos() {
+        Page<Book> bookPage = new PageImpl<>(List.of(book, book));
+        Pageable defaultPageable = PageRequest.of(0, 20);
+        when(bookRepository.findAll(defaultPageable)).thenReturn(bookPage);
+        when(bookMapper.toDtoWithoutCategoryIds(any(Book.class))).thenReturn(
+                toBookDtoWithoutCategoryIds(book));
+
+        List<BookDtoWithoutCategoryIds> response = bookService.findAll(defaultPageable);
+
+        assertEquals(bookPage.getSize(), response.size());
+        verify(bookRepository).findAll(defaultPageable);
+        verify(bookMapper, times(bookPage.getSize())).toDtoWithoutCategoryIds(any(Book.class));
+        verifyNoMoreInteractions(bookRepository, bookMapper);
+    }
+
+    @Test
+    @DisplayName("Get existing book by ID")
+    void getBookDtoById_ExistingId_ReturnsBookDto() {
+        BookDto expected = new BookDto();
```

Comment:
seems to me it's better to create a private method for creating BookDto because it repeats

Reply:
agree